patches/api/0420-Brigadier-based-command-API.patch

```diff
@@ -0,0 +1,746 @@
++
++import com.destroystokyo.paper.profile.PlayerProfile;
++import com.mojang.brigadier.arguments.ArgumentType;
++import io.papermc.paper.math.BlockPosition;
++import net.kyori.adventure.text.Component;
++import org.bukkit.Color;
++import org.bukkit.block.BlockState;
++import org.bukkit.entity.Entity;
++import org.bukkit.entity.Player;
++import org.bukkit.inventory.ItemStack;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.NotNull;
++
++import java.util.Collection;
++import java.util.ServiceLoader;
++import java.util.UUID;
++
++/**
++ * Vanilla argument registry
++ * <p>
++ * These provide rich argument parsing on the client and
++ * may also provide additional signing context.
++ */
++@ApiStatus.Experimental
++public final class VanillaArguments {
++
++    @NotNull
++    static final VanillaArgumentProvider PROVIDER = ServiceLoader.load(VanillaArgumentProvider.class)
++        .findFirst()
++        .orElseThrow();
++
++    /**
++     * An ItemStack argument which provides rich parsing for
++     * specifying item material and item NBT information.
++     *
++     * @return argument
++     */
++    @NotNull
++    public static ArgumentType<ItemStack> itemStack() {
++        return PROVIDER.itemStack();
++    }
++
++    /**
++     * A blockstate argument which will provide rich parsing for specifying
++     * the specific block variant and then the block entity NBT if applicable.
++     *
++     * @return argument
++     */
++    @NotNull
++    public static ArgumentType<BlockState> blockState() {
++        return PROVIDER.blockState();
++    }
++
++    /**
++     * A uuid argument.
++     *
++     * @return argument
++     */
++    @NotNull
++    public static ArgumentType<UUID> uuid() {
++        return PROVIDER.uuid();
++    }
++
++    /**
++     * Represents a selector that can capture any
++     * entity.
++     *
++     * @return argument that takes one entity
++     */
++    @NotNull
++    public static ArgumentType<ArgumentResolver<Entity>> entity() {
++        return PROVIDER.entity();
++    }
++
++    /**
++     * Represents a selector that can capture multiple
++     * entities.
++     *
++     * @return argument that takes multiple entities
++     */
++    @NotNull
++    public static ArgumentType<ArgumentResolver<Collection<Entity>>> entities() {
++        return PROVIDER.entities();
++    }
++
++    /**
++     * Represents a selector that can capture a
++     * player entity.
++     *
++     * @return argument that takes one player
++     */
++    @NotNull
++    public static ArgumentType<ArgumentResolver<Player>> player() {
++        return PROVIDER.player();
++    }
++
++    /**
++     * Represents a selector that can capture multiple
++     * player entities.
++     *
++     * @return argument that takes multiple players
++     */
++    @NotNull
++    public static ArgumentType<ArgumentResolver<Collection<Player>>> players() {
++        return PROVIDER.players();
++    }
```

Comment:
Having the return type not have any "selector" context I feel like limits any attempts to expand on the EntitySelector stuff. Like you can't see if the search was world-limited, or there was a count limit, or a bunch of other stuff that might be useful to expose in the future. Maybe add a subtype of `ArgumentResolver` like `EntitySelectorResolver` that we could add that stuff to in the future.